packages/react-native-web/src/exports/Modal/index.js

```diff
@@ -1,2 +1,147 @@
-import UnimplementedView from '../../modules/UnimplementedView';
-export default UnimplementedView;
+/**
+ * Copyright (c) Nicolas Gallagher.
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React, { forwardRef, useCallback, useMemo, useEffect } from 'react';
+
+import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';
+
+import View from '../View';
+import StyleSheet from '../StyleSheet';
+
+import type { ModalProps } from './types';
+
+import ModalPortal from './ModalPortal';
+import ModalAnimation from './ModalAnimation';
+import ModalFocusTrap from './ModalFocusTrap';
+
+let uniqueModalIdentifier = 0;
+
+const visibleModalStack = [];
+
+function isTopModal(modalId) {
+  if (visibleModalStack.length === 0) {
+    return false;
+  }
+
+  return visibleModalStack[visibleModalStack.length - 1] === modalId;
+}
+
+const Modal = forwardRef<ModalProps, *>((props, forwardedRef) => {
+  const {
+    visible,
+    animationType,
+    transparent,
+    children,
+    onShow,
+    onDismiss,
+    onRequestClose
+  } = props;
+
+  // Set a unique model identifier so we can correctly route
+  // dismissals and check the layering of modals.
+  const modalId = useMemo(() => uniqueModalIdentifier++, []);
+
+  const onDismissCallback = useCallback(() => {
+    // When we dismiss we can't assume that we're dismissing the
+    // top element in the stack - so search the stack and remove
+    // ourselves from it if need be.
+    if (visibleModalStack.indexOf(modalId) !== -1) {
+      visibleModalStack.splice(visibleModalStack.indexOf(modalId), 1);
```

Comment:
If the modal just stops existing - eg, it's removed from the render tree - this won't fire and we end up with the modal stack in a bad spot.

We should probably use an effect to remove the element from the stack in the case of the modal being destroyed

```
useEffect(() => () => removeModal(modalId), [modalId])
```